src/passes/I64ToI32Lowering.cpp

```diff
@@ -696,6 +696,38 @@ struct I64ToI32Lowering : public WalkerPass<PostWalker<I64ToI32Lowering>> {
     // Mostly just shuffling things around here with coercions and whatnot!
     // Note though that all arithmetic is done with f64 to have as much
     // precision as we can.
+    //
+    // NB: this is *not* accurate for i64 -> f32. Using f64s for intermediate
+    // operations can give slightly inaccurate results in some cases, as we
+    // round to an f64, then round again to an f32, which is not always the
+    // same as a single rounding of i64 to f32 directly. Example:
+    //
+    //   #include <stdio.h>
+    //   int main() {
+    //     unsigned long long x = 18446743523953737727ULL;
+    //     float y = x;
+    //     double z = x;
+    //     float w = z;
+    //     printf("i64          : %llu\n"
+    //            "i64->f32     : %f\n"
+    //            "i64->f64     : %f\n"
+    //            "i64->f64->f32: %f\n", x, y, z, w);
+    //   }
+    //
+    //   i64          : 18446743523953737727
+    //   i64->f32     : 18446742974197923840.000000 ;; correct rounding to f32
+    //   i64->f64     : 18446743523953737728.000000 ;; correct rounding to f64
+    //   i64->f64->f32: 18446744073709551616.000000 ;; incorrect rounding to f32
+    //
+    // This is even a problem if we use BigInts in JavaScript to represent
+    // i64s, as Math.fround(BigInt) is not supported - the BigInt must be
+    // converted to a Number first, so we again have that extra rounding.
+    //
+    // A more precise approach could use compiled floatdisf/floatundisf from
+    // compiler-rt, but that is much larger and slower. (Note that we are in the
+    // interesting situation of having f32 and f64 operations and only missing
+    // i64 ones, so we have a different problem to solve than compiler-rt, and
+    // maybe there is a better solution we haven't found yet.)
```

Comment:
Fascinating. Looks like this is an effect from first rounding to an unsafe value outside of `f64`'s safe integer range, with the just introduced error making the subsequent rounding to `f32` imprecise. In the case of `18446743523953737727`, the error of the `i64->f64` operation (that we can compute in JS) is `1`.

```js
var expected = 18446743523953737727n;
var actual = BigInt(Number(expected));
var error = actual - expected;
console.log("expected", expected, "actual", actual, "error", error);
// expected 18446743523953737727n actual 18446743523953737728n error 1n
```

So what happens there is that we are actually rounding `18446743523953737728` to `f32`, not `18446743523953737727`.

```
i64          : 18446743523953737728
i64->f32     : 18446744073709551616.000000
i64->f64     : 18446743523953737728.000000
i64->f64->f32: 18446744073709551616.000000
```

To be continued, maybe.

Reply:
Interesting :)

```js
var expected = 18446743523953737727n;
var actual = Number(expected);
var error = BigInt(actual) - expected;
if (error) {
  let f = new Float32Array([actual]);
  let u = new Uint32Array(f.buffer);
  u[0] += error > 0n ? -1 : 1; // threshold?
  actual = f[0];
}
console.log(BigInt(actual));
// 18446742974197923840n
```

Reply:
Based on that, one strategy may be to do

```
i64tof32(theI64) {
  theF64 = (f64)theI64
  if (theI64 == (i64)theF64) {
    return (f32)theF64
  }
  return floatdisf(theI64)
}
```

to make common cases relatively fast and be correct otherwise :)

Reply:
That's fast solution. I'm not sure it works for all cases but for that example works as expected. On JS:
```js
function floatdisf_emulated(x) {
  let u = x < 0 ? -x : x;
  let f = Math.fround(Number(u >> 32n)) * Number(1n << 32n) + Number(u & 0xFFFFFFFFn);
  return Math.fround(x < 0 ? -f : f);
}

console.log(
  floatdisf_emulated(18446743523953737727n), 
  Math.fround(18446742974197923840.0)
);
```

Print:
```
18446742974197924000 18446742974197924000
```

Reply:
It might be worth finding all the "bad" cases here - there may be a limited number. We can then special-case them perhaps, and avoid floatdisf entirely.

Reply:
Yeah, after deeper testing it not solve problem unfortunately

Reply:
The bad cases are `theIorU64 > 9007199254740991 || theI64 < -9007199254740991` as far as I understand, i.e. `Number.MAX_/MIN_SAFE_INTEGER`.

Reply:
Yeah, it's just a small fraction of numbers there, I think... maybe there's a pattern. If there is, we may be able to add a manual correction (round up or down by 1) specifically for them.

Reply:
I don't have a proper explanation yet, but here's what I know.

The first few bad cases have the following patterns:

```
INCORRECT 9007199791611905
i64->f32      : 9007200328482816.000000 01011010 00000000 00000000 00000001
i64->f64->f32 : 9007199254740992.000000 01011010 00000000 00000000 00000000
i64->f64      : 9007199791611904.000000
hex   : 20000020000001

INCORRECT 9007200865353727
i64->f32      : 9007200328482816.000000 01011010 00000000 00000000 00000001
i64->f64->f32 : 9007201402224640.000000 01011010 00000000 00000000 00000010
i64->f64      : 9007200865353728.000000
hex   : 002000005fffffff

INCORRECT 9007201939095553
i64->f32      : 9007202475966464.000000
i64->f64->f32 : 9007201402224640.000000
i64->f64      : 9007201939095552.000000
hex   : 200000a0000001

INCORRECT 9007203012837375
i64->f32      : 9007202475966464.000000
i64->f64->f32 : 9007203549708288.000000
i64->f64      : 9007203012837376.000000
hex   : 200000dfffffff
```

repeating. The last bad one is

```
INCORRECT 18446743523953737727
i64->f32      : 18446742974197923840.000000  1011111 01111111 11111111 11111111
i64->f64->f32 : 18446744073709551616.000000 01011111 10000000 00000000 00000000
i64->f64      : 18446743523953737728.000000
hex   : ffffff7fffffffff
```

which exhibits a different pattern (overflows into the exponent). In between those, it appears that the ranges increase, indicating that the initial pattern shifts left (not 100% sure). Need to look at those in between.

Reply:
A little hybrid using only relevant branches from `floatdisf`:

```ts
export function i64tof32_asm(value: i64): f32 {
  return a < 0
    ? -u64tof32_asm(-a)
    : u64tof32_asm(a);
}

export function u64tof32_asm(a: u64) : f32 {
  if (a <= (1 << 53)) {
    return <f32><f64>a;
  }
  const FLT_MANT_DIG = 24;
  let sd = 64 - clz(a); // 54..64
  let e = <u32>sd - 1;
  a = (a >> (sd - (FLT_MANT_DIG + 2))) |
      <u64>((a & (~1 >>> ((64 + FLT_MANT_DIG + 2) - sd))) != 0);
  a |= (a & 4) >> 2;
  ++a;
  a >>= 2;
  if (a & (1 << FLT_MANT_DIG)) {
    a >>= 1;
    ++e;
  }
  return reinterpret<f32>(
    (e + 127) << 23 |
    (<u32>a & 0x007FFFFF)
  );
}
```

Reply:
Repo to play with it: https://github.com/dcodeIO/wasm2js_i64tof32